stackrox/stackrox
# ROX-21178: Add buffer in the runtime detection path

## Description

Runtime detection (process indicators and network flows) are now sent to internal queues where the events are buffered after enhancement. The detector pulls from these internal queues and sends these events to the runtime detector.

- If the FF `ROX_CAPTURE_INTERMEDIATE_EVENTS` is enabled and Sensor is offline, these internal queues will be paused. This means no events can be pulled from them.
- If the queues are full, any new event will be dropped. This prevents Sensor from OOMing if we are offline.
- Metrics for the queues' size and number of events dropped have been added.
- The sizes are configurable with `ROX_SENSOR_DETECTOR_PROCESS_INDICATOR_BUFFER_SIZE` and `ROX_SENSOR_DETECTOR_NETWORK_FLOW_BUFFER_SIZE`

## Checklist

If any of these don't apply, please comment below.

## Testing Performed

- [x] The pausable queue is covered with unit tests.
  - Deploy ACS
  - Create a system policy that checks for the execution of a specific command (e.g. `curl`)
  - Break the connection between Central and Sensor
  - Create a pod and run some (`curl`) commands: `kubectl -n test run talk --rm -it --image alpine -- /bin/sh`
  - Re-establish the connection between Central and Sensor
  - Observe the violations that appear in the UI

### Reminder for reviewers

In addition to reviewing code here, reviewers **must** also review testing and request further testing in case the
performed one does not seem sufficient. As a reviewer, you must not approve the change until you understand the
performed testing and you are satisfied with it.
